sdk/eventgrid/eventgrid/review/eventgrid.api.md

```diff
@@ -0,0 +1,206 @@
+## API Report File for "@azure/eventgrid"
+
+> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).
+
+```ts
+
+import { AzureKeyCredential } from '@azure/core-auth';
+import { KeyCredential } from '@azure/core-auth';
+import { OperationOptions } from '@azure/core-http';
+import { PipelineOptions } from '@azure/core-http';
+import { RestResponse } from '@azure/core-http';
+
+export { AzureKeyCredential }
+
+// @public
+export interface CloudEvent<T> {
+    [keyof: string]: any;
+    data?: T;
+    datacontenttype?: string;
+    dataschema?: string;
+    id: string;
+    source: string;
+    subject?: string;
+    time?: Date;
+    type: string;
+}
+
+// @public
+export interface ContainerRegistryArtifactEventData {
+    action?: string;
+    id?: string;
+    target?: ContainerRegistryArtifactEventTarget;
+    timestamp?: Date;
+}
+
+// @public
+export interface ContainerRegistryArtifactEventTarget {
+    digest?: string;
+    mediaType?: string;
+    name?: string;
+    repository?: string;
+    size?: number;
+    tag?: string;
+    version?: string;
+}
+
+// @public
+export type ContainerRegistryChartDeletedEventData = ContainerRegistryArtifactEventData & {};
+
+// @public
+export type ContainerRegistryChartPushedEventData = ContainerRegistryArtifactEventData & {};
+
+// @public
+export interface ContainerRegistryEventActor {
+    name?: string;
+}
+
+// @public
+export interface ContainerRegistryEventData {
+    action?: string;
+    actor?: ContainerRegistryEventActor;
+    id?: string;
+    request?: ContainerRegistryEventRequest;
+    source?: ContainerRegistryEventSource;
+    target?: ContainerRegistryEventTarget;
+    timestamp?: Date;
+}
+
+// @public
+export interface ContainerRegistryEventRequest {
+    addr?: string;
+    host?: string;
+    id?: string;
+    method?: string;
+    useragent?: string;
+}
+
+// @public
+export interface ContainerRegistryEventSource {
+    addr?: string;
+    instanceID?: string;
+}
+
+// @public
+export interface ContainerRegistryEventTarget {
+    digest?: string;
+    length?: number;
+    mediaType?: string;
+    repository?: string;
+    size?: number;
+    tag?: string;
+    url?: string;
+}
+
+// @public
+export type ContainerRegistryImageDeletedEventData = ContainerRegistryEventData & {};
+
+// @public
+export type ContainerRegistryImagePushedEventData = ContainerRegistryEventData & {};
+
+// @public
+export class EventGridClient {
+    constructor(endpointUrl: string, credential: KeyCredential | SignatureCredential, options?: EventGridClientOptions);
+    readonly apiVersion: string;
+    readonly endpointUrl: string;
+    generateSharedAccessSignature(expiresOnUtc: Date): string;
+    sendCloudEvents(events: CloudEvent<any>[], options?: SendCloudEventsOptions): Promise<RestResponse>;
```

Comment:
Is there a strong reason to split up sending events into three methods? You could overload a single `send` method instead perhaps.

Reply:
> Is there a strong reason to split up sending events into three methods?

The underlying service call differs slightly between sending Cloud Events and sending either Custom Events or Event Grid Events.

When sending Cloud Events, we have to use a special `Content-Type` of `application/cloudevents+json` instead of `application/json`.  On the generated service client we have separate overloads for these cases.

The problem with a single method with overloads is we need a way to if the data passed in was a CloudEvent or not, so we can make the right call to the service.  But because we are using interfaces and not class types, there's no good way to do this.  We could probe to see if the object passed in has a `source` property (which is required for cloud events, but not allowed for events encoded in the event grid schema) and go from there.  However, the Event Grid service also supports "Custom Schema Events", which allows a customer to create an Event Grid topic that expects events in user defined envelope (when creating the resource, you define what the relevant property names are and how they map to properties that the event grid service expects, then you can send your events with that custom envelope format directly to the service).

Because of the above, there isn't a good way to use the existence of certain properties to distinguish between an event encoded in the cloud event schema and an event using a custom schema.

We could use the existence of the `source` property on the input to decide if it meant cloud events or not, and then provide an option in `SendCloudEventsOptions` which would let you force sending it as a custom event schema.  The one possible issue here is that we actually have different generated client method for an event in a custom schema (to ensure that all the fields of the object are serialized instead of just the "known" ones as in the case of an event in the Event Grid schema), but maybe it could be made to work?

We could also consider just dropping support for sending events in the custom event schema.  The Track 1 library did not support this and the service team may be fine with not having SDK support for this feature (in general the landscape is moving towards Cloud Events).

Reply:
In the previous board discussion about this I was convinced at least that having separate methods is preferable because it's less ambiguous how to handle the provided data correctly. Also it doesn't seem like many folks would want to mix/match?

Reply:
> In the previous board discussion about this I was convinced at least that having separate methods is preferable because it's less ambiguous how to handle the provided data correctly. Also it doesn't seem like many folks would want to mix/match?

Regarding mixing and matching, you are correct.  When you create a topic, you specify the schema it should expect, and you MUST send events to it in that schema.  Having the schema type elevated into the method name felt good to me because maybe you're more likely to say "Hey, I know I have a topic that expects cloud events, so it's a code smell here that I'm calling some method other than `sendCloudEvents`".

FWIW - I think that the other languages are adopting a pattern of having a single method on the client, since they are able to do richer type tests at runtime to figure out what state your in (or in the case of .NET and Java, the complier does it for you during overload selection).

I would also be fine with something like this (it's a little wacky, but again, I think I'm okay with some wackiness for custom events):

```typescript

export class CustomEvent {
    constructor(event: Record<string, any>); // stores the event in an `@internal` field)
}

export class EventGridClient {
  // other members elided)
  send(events: EventGridEvent<any>[] | CloudEvent<any>[] | CustomEvent[]);
}
```

We would, of course, follow the overload patterns to provide good doc comments for each version.  In this world, I could simply do an `instanceof` check to determine if something is a custom event, and if it's not, use property sniffing to determine if we were in the EventGridEvent case or the CloudEvent case.

This would let you continue to use object literals for both EventGridEvents and CloudEvents, and for custom events, you'd have a bit of ceremony in that you have to do something like:

```typescript
client.send([
  new CustomEvent({ /* some object literal that represents your custom event })
]);
```

Maybe that's not the end of the world?  I know in general we sort of frown on uses classes, which is one of the reasons I didn't go down this path.  It is, however, similar to what Python will be doing.

Reply:
I prefer to avoid `instanceof` and class wrappers for objects-as-data, both of which to me are more upsetting than a couple extra members.

I think if we **must** collapse to one method, I'd prefer to actually make the choice explicit in the signature and do overloads based on the discriminant much like event listeners e.g.

```ts
export class EventGridClient {
  send("EventGridEvent", events: EventGridEvent<any>[]);
  send("CloudEvent", events: CloudEvent<any>[]);
  send("CustomEvent", events: CustomEvent[]);
}
```

That way it's still one method, but you have to actually tell us what kind of event you're giving us and TS can always easily disambiguate the argument list and provide good reference docs.

Reply:
Per our discussion during the arch board review, we are ok with sticking with the pattern we currently have (no overloads and the "schema" is encoded in the name of the function)